docs/proposals/self-service-notifications.md

```diff
@@ -0,0 +1,144 @@
+This function uses the application's `metadata.namespace` field to find the location of the configmap and secret. 
+This controller uses both the self-service configuration and the admin defined default configuration.
+
+![img.png](images/self-service-notifications.png)
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+#### Use case 1: application team wants to configure ArgoCD notification using pager duty V2.
+I want to receive Pager Duty notification when my application is syncing, but our default ArgoCD notification is using slack.
+
+* Create two additional resources `argocd-notifications-secret` and `argocd-notifications-cm`.
+  In these resources I used PagerDutyV2 as service type.
+* Deploy these two additional resources to the same namespace as my ArgoCD application.
+
+
+#### Use case 1 Detailed example
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  annotations:
+    notifications.argoproj.io/subscribe.on-sync-running.pagerdutyv2: hello-2_xxxx
+    notifications.argoproj.io/subscribe.on-sync-running.slack: may-test
+  name: guest-book
+  namespace: app-may-test
+spec:
+  destination:
+    namespace: app-may-test
+    server: https://xxxx
+  project: default
+  source:
+    path: guestbook
+    repoURL: https://github.com/mayzhang2000/argocd-example-apps.git
+    targetRevision: HEAD
+```
+
+```yaml
+apiVersion: v1
+data:
+  service.pagerdutyv2: |
+    serviceKeys:
+      hello-2_xxxxx: $pagerdutyv2-key-hello-2_xxxx
+  template.app-sync-running: |
+    pagerdutyv2:
+      summary: "App {{.app.metadata.name}} sync running "
+      severity: "info"
+      source: "{{.app.metadata.name}}"
+  trigger.on-sync-running: |
+    - description: Application is being synced
+      send:
+      - app-sync-running
+      when: app.status.operationState.phase in ['Running']
+kind: ConfigMap
+metadata:
+  labels:
+    app.kubernetes.io/instance: mayeast
+  name: argocd-notifications-cm
+  namespace: app-may-test
+```
+
+```yaml
+apiVersion: v1
+data:
+  pagerdutyv2-key-hello-2_4759196499290493255: ++++++++
+kind: Secret
+metadata:
+  labels:
+    app.kubernetes.io/instance: mayeast
+  name: argocd-notifications-secret
+  namespace: app-may-test
+type: Opaque
+```
+
+#### Use case 2: application team wants to configure custom template
+ArgoCD admin has configured basic templates, as end user, I want to receive more detailed information that only pertains to my team.
+
+#### Use case 3: application team wants to configure custom trigger
+ArgoCD admin has configured basic triggers, as end user, I want to receive a notification based on a trigger that pertains to my team.
+
+### Security Considerations
+
+### Risks and Mitigations
+
+### Considerations
+This adds a new configuration pattern to ArgoCD. Today, ArgoCD has a few different patterns for configuration. 
+This proposal aligns itself with apps in any namespace pattern.
+
+## Drawbacks
+
+## Alternatives
```

Comment:
```suggestion
## Alternatives

Multitenancy-by-namespace is a design decision which particularly suits gitops. Argo CD also supports multitenancy governed by AppProjects. But that multitenancy model must be completely managed by the Argo CD admin, and users may not self-serve Argo CD config via GitOps.

Alternatives to this gitops-first design could include:
1) Allowing self-serve notifications config via the API, segmented by project (i.e. not gitops)
2) Associating notifications config with an AppProject instead of an Applications namespace (could still be gitops, depending on the association mechanism)
3) Associating notifications config with an Application directly (maybe via a ConfigMapRef field or an annotation?) instead of an Applications namespace (could probably also be gitopsed, but is a bit more complicated)

This choice is important, because it sets a precedent for how we expect self-service, gitops-first multi-tenancy to scale in Argo. Basically: how do we let users configure _stuff_ about their applications via gitops without requiring massive administrative overhead. The model we establish could later expand to configuring resource customizations, sync windows, etc.
```